pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,350 @@
+			err = verifyPodImageWithHostPathVolume(cs, nodeName, CalicoUpgradeDir)
+			if err != nil {
+				log.WithError(err).Fatal("Failed to verify windows-upgrade pod image")
+			}
+
+			err = uninstall()
+			if err != nil {
+				log.WithError(err).Error("Uninstall failed, will retry")
+				break
+			}
+
+			time.Sleep(3 * time.Second)
+			err = execScript(script)
+			if err != nil {
+				log.WithError(err).Fatal("Failed to upgrade to new version")
+			}
+
+			// Upgrade will run in another process. The running
+			// calico-upgrade service is done. The new calico-upgrade
+			// service will clean the old service up.
+			date := time.Now().Format("2006-01-02")
+			log.Info(fmt.Sprintf("Upgrade is in progress. Upgrade log is in c:\\calico-upgrade.%v.log", date))
+			time.Sleep(3 * time.Second)
+			return
+		}
+	}
+}
+
+// Return the base directory for Calico upgrade service.
+func baseDir() string {
+	dir := filepath.Dir(os.Args[0])
+	return "c:\\" + filepath.Base(dir)
+}
+
+// Return if the monitor service is running as part of Enterprise installation.
+func runningEnterprise() bool {
+	return baseDir() == EnterpriseBaseDir
+}
+
+// Return kubeconfig file path for Calico
+func kubeConfigFile() string {
+	return baseDir() + "\\" + CalicoKubeConfigFile
+}
+
+func uninstall() error {
+	path := filepath.Join(baseDir(), "uninstall-calico.ps1")
+	log.Infof("Start uninstall script %s\n", path)
+	stdout, stderr, err := powershell(path)
+	fmt.Println(stdout, stderr)
+	if err != nil {
+		return err
+	}
+	// After the uninstall completes, move the existing calico-node.exe to
+	// a temporary file. The calico-upgrade service is still running so not
+	// doing this means we cannot replace calico-node.exe with the upgrade.
+	stdout, stderr, err = powershell(fmt.Sprintf(`mv %v\calico-node.exe %v\calico-node.exe.to-be-replaced`, baseDir(), baseDir()))
+	fmt.Println(stdout, stderr)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func execScript(script string) error {
+	log.Infof("Start script %s\n", script)
+
+	// This has to be done in a separate process because when the new calico services are started, the existing
+	// calico-upgrade service is removed so the new calico-upgrade service can be started.
+	// However, removing the existing calico-upgrade service means the powershell
+	// process running the upgrade script is killed and the installation is left
+	// incomplete.
+	cmd := fmt.Sprintf(`Start-Process powershell -argument %q -WindowStyle hidden`, script)
+	stdout, stderr, err := powershell(cmd)
+
+	if err != nil {
+		return err
+	}
+	fmt.Println(stdout, stderr)
+	return nil
+}
+
+func verifyImagesSharePathPrefix(first, second string) error {
+	n1, err := image.ParseNamed(first)
+	if err != nil {
+		return err
+	}
+	n2, err := image.ParseNamed(second)
+	if err != nil {
+		return err
+	}
+	// Compare the domain parts (e.g. docker.io) of the image references.
+	if image.Domain(n1) != image.Domain(n2) {
```

Comment:
Are we sure if image.Domain(n1) is not ""?

Reply:
Yes, Domain will always be defined, reference.ParseNamed fails without a valid domain in the image reference.